# Justin Elkinson
### Worked with Cam Germaine

### Train and tune predictive models

In [1]:
import pandas as pd
import numpy as np
from sklearn import tree
from sklearn.tree import DecisionTreeClassifier
from sklearn import metrics

In [2]:
# Import the data
df = pd.read_csv('BC Data.csv')

# Assign input variables
X = df.loc[:,['radius_mean','texture_mean', 'perimeter_mean', 'area_mean', 'smoothness_mean', 'compactness_mean', 'concavity_mean', 'concave points_mean',
             'symmetry_mean', 'fractal_dimension_mean',
             'radius_se','texture_se', 'perimeter_se', 'area_se', 'smoothness_se', 'compactness_se', 'concavity_se', 'concave points_se',
             'symmetry_se', 'fractal_dimension_se']]

# Assign target variable
y = df['diagnosis']


In [3]:
from sklearn.model_selection import train_test_split

x_train,x_test,y_train,y_test=train_test_split(X,y,test_size=0.2)

In [4]:
# Create a decision tree and train

tree1 = DecisionTreeClassifier()
tree1.fit(x_train, y_train)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
                       max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=None, splitter='best')

In [5]:
# Accuracry on training data
tree1.score(x_train, y_train)

1.0

In [6]:
# Prediction and accuracy on testing data
y_pred = tree1.predict(x_test)
e1 = metrics.accuracy_score(y_pred, y_test)
e1

0.8947368421052632

In [7]:
# Create another decision tree and train
tree2 = DecisionTreeClassifier(criterion='entropy', max_depth=3)
tree2.fit(x_train, y_train)

DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=3,
                       max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=None, splitter='best')

In [8]:
tree2.score(x_train, y_train)
y_pred2 = tree2.predict(x_test)
e2=metrics.accuracy_score(y_pred2, y_test)
e2

0.8859649122807017

#### The accuracy of the decision tree with max-depth 3 and entropy criterion is 88.6%

In [9]:
### Visualize the tree
p=tree.plot_tree(tree2)

### TUNE THE MODEL:

In [10]:
# Decide what hyperparameter to tune then decide the searching range
param_grid = {'max_depth': range(2,7), 'criterion':['gini', 'entropy']}

# Create a list of trees
from sklearn.model_selection import GridSearchCV
tree3 = GridSearchCV(DecisionTreeClassifier(), param_grid, cv = 5)

In [11]:
tree3.fit(x_train, y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
             estimator=DecisionTreeClassifier(class_weight=None,
                                              criterion='gini', max_depth=None,
                                              max_features=None,
                                              max_leaf_nodes=None,
                                              min_impurity_decrease=0.0,
                                              min_impurity_split=None,
                                              min_samples_leaf=1,
                                              min_samples_split=2,
                                              min_weight_fraction_leaf=0.0,
                                              presort=False, random_state=None,
                                              splitter='best'),
             iid='warn', n_jobs=None,
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': range(2, 7)},
             pre_dispatch

In [12]:
# Show the best found paramters
tree3.best_params_

{'criterion': 'gini', 'max_depth': 3}

In [16]:
y_pred2 = tree3.predict(x_test)
e3=metrics.accuracy_score(y_pred2, y_test)
e3

0.8771929824561403

#### Best hyperparameter criterion = gini and max depth = 3
#### test data accuracy = 87.72%

### Comparing the three trees:

In [14]:
[e1, e2, e3]

[0.8947368421052632, 0.8859649122807017, 0.8771929824561403]